setup.py

```diff
@@ -35,70 +35,49 @@ def is_arm():
-    return ''
+        return ['-DCMAKE_C_FLAGS=-m32']
+    return []
 
 
-def determine_generator_string():
+def determine_generator_args():
     if sys.platform == 'win32':
-        prog_x86_path = os.getenv('PROGRAMFILES(x86)')
-        if os.path.exists(prog_x86_path + '\\Microsoft Visual Studio\\2019'):
-            vs_version = '16.0'
-            print('found installed version of Visual Studio 2019')
-        elif os.path.exists(prog_x86_path + '\\Microsoft Visual Studio\\2017'):
-            vs_version = '15.0'
-            print('found installed version of Visual Studio 2017')
-        elif os.path.exists(prog_x86_path + '\\Microsoft Visual Studio 14.0'):
-            vs_version = '14.0'
-            print('found installed version of Visual Studio 2015')
-        else:
-            print('Making an attempt at calling vswhere')
-            vswhere_args = [
-                '%ProgramFiles(x86)%\\Microsoft Visual Studio\\Installer\\vswhere.exe',
-                '-legacy',
-                '-latest',
-                '-property',
-                'installationVersion']
-            vswhere_output = None
-
-            try:
-                vswhere_output = subprocess.check_output(vswhere_args, shell=True)
-            except subprocess.CalledProcessError:
-                raise RuntimeError('No version of MSVC compiler could be found!')
-
-            if vswhere_output is not None:
-                for out in vswhere_output.split():
-                    vs_version = out.decode('utf-8')
-            else:
-                raise RuntimeError('No MSVC compiler could be found!')
-
-        vs_major_version = vs_version.split('.')[0]
-
-        cmake_list_gen_args = ['cmake', '--help']
-        cmake_help_output = subprocess.check_output(cmake_list_gen_args)
-
-        vs_version_gen_str = None
-        for out in cmake_help_output.splitlines():
-            trimmed_out = out.decode('utf-8').strip()
-            if 'Visual Studio' in trimmed_out and vs_major_version in trimmed_out:
-                print('selecting generator {}'.format(trimmed_out))
-                vs_version_gen_str = trimmed_out.split('[')[0].strip(' *')
-                break
-
-        if vs_version_gen_str is None:
-            raise RuntimeError('CMake does not recognize an installed version of visual studio on your system.')
-
-        if is_64bit():
-            print('64bit version of python detected, using win64 builds')
-            vs_version_gen_str = vs_version_gen_str + ' Win64'
-
-        vs_version_gen_str = '-G' + vs_version_gen_str
-        print('Succesfully determined generator as \"{}\"'.format(vs_version_gen_str))
-        return vs_version_gen_str
-    return ''
+        try:
+            # See which compiler python picks
+            compiler = distutils.ccompiler.new_compiler()
+            compiler.initialize()
+
+            # Look at compiler path to divine the Visual Studio version
+            if '\\Microsoft Visual Studio\\2019' in compiler.cc:
```

Comment:
Any worries about customers with non-standard install locations or do we assume that this is just controllable builds for publishing?

Are we assuming that this is being invoked from a dev-tools prompt or anywhere?

Reply:
This works with vanilla cmd.exe or powershell,

You're right about it breaking with custom install paths. This is the best I could come up with without using private library functions. I added a comment in code explaining as much